### DISCUSSION TITLE
Is there a better way to show file name only when do grep?

### DISCUSSION THREAD

**Discussion by cpcopy** [ibhagwan/fzf-lua#1407] [Q&A] [answered]
Currently, I use following configuration. The problem is multline configuration doesn't work anymore and no position information.
```
        grep = {
            multiline = 1,      -- Display as: PATH:LINE:COL\nTEXT
            formatter = "path.filename_first",
            fzf_opts = {
                ["--with-nth"] = '1,2,4..', -- 1: file icon, 2: file name, 3: file path,
            },
        }
```

**Comment by ibhagwan** [comment]
Without writing your own formatter (or a custom picker) you’re limited to playing around with `--nth`, what you could do is maybe set the delimiter to both `utils.nbsp` and  `:` and then you’ll be able to extract the positron info as it’s wrapped with `:`, in any event not optimal.

An example where multiple delimiters are used:
https://github.com/ibhagwan/fzf-lua/blob/253c6ff931eff4edeb9e48095a37e39a818ea492/lua/fzf-lua/defaults.lua#L658

**Reply by cpcopy** [reply]
Thanks. Just know delimiter can be multiple.  But it is still failed if I set:
```
["--delimiter"] = "[: ]"
```
The space between filename and path is an en space? I tried to follow the source code set as below
```
["--delimiter"] = string.format("[:%s]", "\u{2002}")
```
It doesn't work neither. My lua version is 5.1.

**Comment by cpcopy** [comment]
I switched to `grep_project`. It looks like `["--delimiter"]` isn't applied to `grep_project`. It always use : as delimiter.

**Reply by ibhagwan** [reply]
`grep_project` sets the limiter by code so it can ignore the file names and grep only the contents, if you set the delimiter in fzf_opts you can override that.

> The space between filename and path is an en space? I tried to follow the source code set as below

Using `\u{}` had an issue (forgot what it was), I therefore used the below instead:
https://github.com/ibhagwan/fzf-lua/blob/40904abf2bdd784aa789515ce3cd0bc267e5eff4/lua/fzf-lua/utils.lua#L65

You’re most likely using LuaJIT but if you’re only Lua 5.1 you’d also need to translate the `\x` yourself:
https://github.com/ibhagwan/fzf-lua/blob/40904abf2bdd784aa789515ce3cd0bc267e5eff4/lua/fzf-lua/utils.lua#L67-L76

**Reply by cpcopy** [reply]
>  `grep_project` sets the limiter by code so it can ignore the file names and grep only the contents, if you set the delimiter in fzf_opts you can override that. 

Checked the code and found it can only be overwritten by function input parameters. Thanks.
>Using \u{} had an issue (forgot what it was), I therefore used the below instead:

Yes. I saw the comments. I tried what you suggested before. It doesn't work. Now, I write the function input parameters as below. Unfortunately, it doesn't work neither. File name and path name are still in one subset. I checked the unicode of print out ['--delimiter'] = '[: ]'. The space is already an en space. 
If I use the default delimiter configuration(no overwritten from configuration), the file name and file path are correctly separated by space, which only works in `live_grep`, not `grep_project`.
```
local grep_opts = function()
    local rg_opts = "rg_opts = '--type c --column --line-number --no-heading --color=always --smart-case --max-columns=4096 -e',"
    local nbsp = [[\xe2\x80\x82]] -- "\u{2002}"

    if _VERSION and type(_VERSION) == "string" then
        local ver = tonumber(_VERSION:match("%d+.%d+"))
        if ver < 5.2 then
            nbsp = nbsp:gsub("\\x(%x%x)",
            function(x)
                return string.char(tonumber(x, 16))
            end)
        end
    end

    local fzf_opts = string.format("fzf_opts = { ['--delimiter'] = '[:%s]' },", nbsp)
    print(fzf_opts)

    return "{ " .. fzf_opts .. rg_opts .. " }"
end
```

**Comment by cpcopy** [comment, marked as answer]
Just find the code for filename_first formatter. `\t` is used as delimiter between tail and parent.
```
 return tail .. "\t" .. parent %s
```
With configuration as below, I can show file name only.
```
["--delimiter"] = "[:\t]"
```

**Reply by ibhagwan** [reply]
> Just find the code for filename_first formatter. \t is used as delimiter between tail and parent.

Correct, that’s how I reconstruct the path back to the original in the formatter `from`.
